contracts/upgradeable_contracts/amb_erc677_to_erc677/HomeStakeTokenMediator.sol

```diff
@@ -0,0 +1,160 @@
+     * @param _owner token proxy contract address
+     */
+    function transferTokenOwnership(address _owner) external onlyOwner {
+        Ownable(erc677token()).transferOwnership(_owner);
+    }
+
+    /**
+     * @dev Updates address of contract used for handling mint operations,
+     * all subsequent mint operations will be called through this contract
+     * @param _mintHandler address of new contract
+     */
+    function setMintHandler(address _mintHandler) external onlyOwner {
+        require(AddressUtils.isContract(_mintHandler));
+        addressStorage[MINT_HANDLER] = _mintHandler;
+    }
+
+    /**
+     * @dev Retrieves currently used contract for handling mint operations, defaults to token itself
+     * @return address of mint handler contract
+     */
+    function getMintHandler() public view returns (IMintHandler) {
+        return IMintHandler(addressStorage[MINT_HANDLER]);
+    }
+
+    /**
+     * @dev Executes action on the request to deposit tokens relayed from the other network
+     * @param _recipient address of tokens receiver
+     * @param _value amount of bridged tokens
+     */
+    function executeActionOnBridgedTokens(address _recipient, uint256 _value) internal {
+        uint256 value = _value.mul(10**decimalShift());
+        getMintHandler().mint(_recipient, value);
+    }
+
+    /**
+     * @dev Executes action on withdrawal of bridged tokens
+     * @param _token address of token contract
+     * @param _from address of tokens sender
+     * @param _value requsted amount of bridged tokens
+     * @param _data alternative receiver, if specified
+     */
+    function bridgeSpecificActionsOnTokenTransfer(ERC677 _token, address _from, uint256 _value, bytes _data) internal {
+        if (!lock()) {
+            // burn all incoming tokens
+            IBurnableMintableERC677Token(_token).burn(_value);
+
+            if (address(_blockRewardContract()) == address(0)) {
+                // in case if block reward contract is not configured, the fee is not collected
+                passMessage(chooseReceiver(_from, _data), _value);
+            } else {
+                // when block reward contract is defined, the calculated fee is subtracted from the original value
+                uint256 fee = calculateFee(_value);
+                passMessage(chooseReceiver(_from, _data), _value.sub(fee));
+                if (fee > 0) {
+                    // the fee itself is distributed later in the block reward contract
+                    _blockRewardContract().addBridgeTokenRewardReceivers(fee);
+                }
+            }
+        }
+    }
+
+    /**
+     * @dev Executes action on relayed request to fix the failed transfer of tokens
+     * @param _recipient address of tokens receiver
+     * @param _value amount of fixed tokens
+     */
+    function executeActionOnFixedTokens(address _recipient, uint256 _value) internal {
+        getMintHandler().mint(_recipient, _value);
+    }
```

Comment:
I have a question regarding `executeActionOnFixedTokens` function: as far as I understand, it's used to return tokens back to sender if something is wrong when passing a message from bridge contract to mediator contract. I tried to analyze the following case (please correct me if I'm wrong somewhere):

For example, some sender Bob on Home side uses `alternative receiver` feature to send tokens to Alice. He sends their tokens to Home mediator contract, so `bridgeSpecificActionsOnTokenTransfer` is triggered (and the tokens are burnt on Home side) and then the message passed to Home bridge contract. Oracle catches the corresponding event, the bridge validators sign the message and request the Foreign bridge contract to pass the message to Foreign mediator contract, but `_passMessage` returns `false` for some reason.

Then someone calls the `requestFailedMessageFix` on the Foreign mediator contract. The Foreign mediator contract in turn through the Foreign bridge contract asks for `fixFailedMessage` to be called on Home side. The `fixFailedMessage` calls `executeActionOnFixedTokens` of the Home mediator contract which mints back the tokens for Alice (but not for Bob) on the Home side.

Is it intended behavior that in case of `alternative receiver` feature the failed tokens are returned back to Alice, but not to Bob on Home side?